addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/handler/KNXBasicThingHandler.java

```diff
@@ -0,0 +1,682 @@
+                            channel.getUID(), selector.isSlave());
+                }
+            });
+        }
+    }
+
+    private void processDataReceived(KNXBridgeBaseThingHandler bridge, GroupAddress destination, byte[] asdu,
+            String dpt, ChannelUID channelUID, boolean slave) {
+
+        if (dpt != null) {
+
+            if (!bridge.isDPTSupported(dpt)) {
+                logger.warn("DPT {} is not supported by the KNX binding.", dpt);
+                return;
+            }
+
+            Datapoint datapoint = new CommandDP(destination, getThing().getUID().toString(), 0, dpt);
+            Type type = bridge.getType(destination, dpt, asdu);
+
+            if (type != null) {
+                if (slave) {
+                    if (type instanceof Command) {
+                        postCommand(channelUID, (Command) type);
+                    }
+                } else {
+                    if (type instanceof State) {
+                        updateState(channelUID, (State) type);
+                    }
+                }
+
+            } else {
+                final char[] hexCode = "0123456789ABCDEF".toCharArray();
+                StringBuilder sb = new StringBuilder(2 + asdu.length * 2);
+                sb.append("0x");
+                for (byte b : asdu) {
+                    sb.append(hexCode[(b >> 4) & 0xF]);
+                    sb.append(hexCode[(b & 0xF)]);
+                }
+
+                logger.warn(
+                        "Ignoring KNX bus data: couldn't transform to Type (not supported). Destination='{}', datapoint='{}', data='{}'",
+                        destination, datapoint, sb);
+                return;
+            }
+
+        }
+    }
+
+    public void restart() {
+        if (address != null) {
+            getBridgeHandler().restartNetworkDevice(address);
+        }
+    }
+
+    class ReadRunnable implements Runnable {
+
+        private GroupAddress address;
+        private String dpt;
+
+        ReadRunnable(GroupAddress address, String dpt) {
+            this.address = address;
+            this.dpt = dpt;
+        }
+
+        @Override
+        public void run() {
+            if (getThing().getStatus() == ThingStatus.ONLINE && getBridge().getStatus() == ThingStatus.ONLINE) {
+                if (!getBridgeHandler().isDPTSupported(dpt)) {
+                    logger.warn("DPT '{}' is not supported by the KNX binding", dpt);
+                    return;
+                }
+                Datapoint datapoint = new CommandDP(address, getThing().getUID().toString(), 0, dpt);
+                getBridgeHandler().readDatapoint(datapoint, getBridgeHandler().getReadRetriesLimit());
+            }
+        }
+    };
+
+    private void pollDeviceStatus() {
+        try {
+            if (address != null && getBridge().getStatus() == ThingStatus.ONLINE) {
+                logger.debug("Polling the individual address {}", address.toString());
```

Comment:
You don't have to call toString it will be called automatically, but only when the logging-level actually applies.